### DISCUSSION TITLE
Warning: Do not call Hooks inside useEffect(...), useMemo(...), or other built-in Hooks

### DISCUSSION THREAD

**Discussion by robojiannis** [firecmsco/firecms#341] [Q&A] [answered]
Hi, 
sorry, I'm still learing React and the ropes of firecms :)

I'm getting this warning when using an extraAction. 
`Warning: Do not call Hooks inside useEffect(...), useMemo(...), or other built-in Hooks. You can only call Hooks at the top level of your React function. For more information, see https://reactjs.org/link/rules-of-hooks
`

I have this simple function:

```
import React from "react";
import {
    SelectionController,
    useStorageSource
} from "@camberi/firecms";
import { Button } from "@mui/material";

export function PrintDecksAction({ selectionController }: {
  selectionController: SelectionController
}) {
  let storage = useStorageSource();
  const onClick = (event: React.MouseEvent) => {
      console.log (selectionController);

  };
  
  return (
      <Button onClick={onClick} color="primary">
          pdf of selection
      </Button>
  );
}
```

which I import and apply as an extraAction:

```
import { PrintDecksAction } from "./collections/decks";


buildCollection({
                path:"decks",
                exportable: false,
                schema:DecksSchema.schema(),
                extraActions:PrintDecksAction,
                name: "Decks",
                permissions: ({entity, path, user, authController, context }) => {
                  return ({
                      read: true,
                      create: isadmin,
                      update: isadmin,
                      delete: isadmin
                  });
              }
}),
```

I'm sure it's something silly, I just can't figure it out. I'd be grateful for your tips 🙏

**Comment by fgatti675** [comment, marked as answer]
Hi @robojiannis :)
Effectively, you can't use the hooks inside a callback, like the actions one.
The good news is that within the params that the callback receives there is a `context` object that includes all the controllers of the CMS, so you can access the storage source from there.
Hope this helps!

**Reply by robojiannis** [reply]
Gotcha, thanks for the swift reply! I'll give it a go
